core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -1176,6 +1223,9 @@ private[spark] class MapOutputTrackerWorker(conf: SparkConf) extends MapOutputTr
   // instantiate a serializer. See the followup to SPARK-36705 for more details.
   private lazy val fetchMergeResult = Utils.isPushBasedShuffleEnabled(conf, isDriver = false)
 
+  // Exposed for testing
+  val shufflePushMergerLocations = new ConcurrentHashMap[Int, Seq[BlockManagerId]]().asScala
```

Comment:
Agree, good callout. This is definitely a potential issue with indeterminate stages. Thanks.

Reply:
Instead of keeping track of both <shuffleId, shuffleMergeId> for shuffle merger locations, I think we can piggy back on `unregisterAllMapAndMergeOutput` to clear the `mergers` in `ShuffleStatus` similarly in the `updateEpoch` (like this [comment](https://github.com/apache/spark/pull/34122#discussion_r785388799)) we can clear the `shufflePushMergerLocations`. This makes the code much simpler and neat. Let me know if you think otherwise or this has any other issues.